polyfill/test/validStrings.mjs

```diff
@@ -63,6 +63,22 @@ function withCode(productionLike, func) {
   return new FromFunction(func, productionLike);
 }
 
+// Generate a token. If the validation function doesn't throw a SyntaxError,
+// return the token. Otherwise, try again.
+function withEarlyError(productionLike, validatorFunc) {
+  let result = productionLike.generate();
+  try {
+    validatorFunc(result);
+    return result;
+  } catch (e) {
+    if (e instanceof SyntaxError) {
+      return withEarlyError(productionLike, validatorFunc);
+    } else {
+      throw e;
+    }
+  }
+}
```

Comment:
Wouldn't it be more consistent and comprehensible to return an object _with_ a `generate` method rather than the results _of_ a `generate()` invocation?

```suggestion
// Return a productionLike that defers to another productionLike for generation
// but uses a validator function to filter out any result that a validator function
// rejects with SyntaxError.
function withSyntaxConstraints(productionLike, validatorFunc) {
  return { generate: filtered };

  function filtered(data) {
    let dataClone = { ...data };
    let result = productionLike.generate(dataClone);
    try {
      validatorFunc(result);
    } catch (e) {
      if (e instanceof SyntaxError) {
        // Generated result violated a constraint; try again
        // (but allow crashing if the call stack gets too deep).
        return filtered(data);
      } else {
        // Propagate any other error.
        throw e;
      }
    }

    // Copy properties that were modified on the cloned input data.
    for (let key of Reflect.ownKeys(dataClone)) {
      data[key] = dataClone[key];
    }

    return result;
  };
}
```